jaxrs-testing/src/main/java/io/airlift/jaxrs/testing/JaxrsTestingHttpProcessor.java

```diff
@@ -0,0 +1,135 @@
+
+    private boolean trace;
+
+    public JaxrsTestingHttpProcessor(URI baseUri, Object... jaxRsSingletons)
+    {
+        Set<Object> jaxRsSingletonsSet = ImmutableSet.copyOf(jaxRsSingletons);
+        Application application = new Application() {
+            @Override
+            public Set<Object> getSingletons()
+            {
+                return jaxRsSingletonsSet;
+            }
+        };
+        TestContainer testContainer = new InMemoryTestContainerFactory()
+                .create(baseUri, DeploymentContext.newInstance(application));
+        ClientConfig clientConfig = testContainer.getClientConfig();
+        this.client = JerseyClientBuilder.createClient(clientConfig);
+    }
+
+    public JaxrsTestingHttpProcessor setTrace(boolean enabled)
+    {
+        this.trace = enabled;
+        return this;
+    }
+
+    @Override
+    public Response handle(Request request)
+            throws Exception
+    {
+        // prepare request to jax-rs resource
+        MultivaluedMap<String, Object> requestHeaders = new MultivaluedHashMap<>();
+        for (Map.Entry<String, String> entry : request.getHeaders().entries()) {
+            requestHeaders.add(entry.getKey(), entry.getValue());
+        }
+        Invocation.Builder invocationBuilder = client.target(request.getUri()).request().headers(requestHeaders);
+        Invocation invocation;
+        if (request.getBodyGenerator() == null) {
+            invocation = invocationBuilder.build(request.getMethod());
+        }
+        else {
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            request.getBodyGenerator().write(byteArrayOutputStream);
+            byteArrayOutputStream.close();
+            byte[] bytes = byteArrayOutputStream.toByteArray();
+            Entity<byte[]> entity = Entity.entity(bytes, (String) getOnlyElement(requestHeaders.get("Content-Type")));
+            invocation = invocationBuilder.build(request.getMethod(), entity);
+        }
+
+        // issue request, and handle exceptions
+        javax.ws.rs.core.Response result;
+        try {
+            result = invocation.invoke(javax.ws.rs.core.Response.class);
+        }
+        catch (ProcessingException exception) {
+            // Peel out the exception we threw in jax-rs resource implementation
+            // to facilitate testing exceptional paths
+            if (trace) {
+                log.warn(exception.getCause(), "%-8s %s -> Exception", request.getMethod(), request.getUri());
+            }
+            throw (Exception) exception.getCause();
+        }
+        catch (Throwable throwable) {
+            if (trace) {
+                log.warn(throwable, "%-8s %s -> Fail", request.getMethod(), request.getUri());
+            }
+            throw throwable;
+        }
+
+        // process response from jax-rs resource
+        ImmutableListMultimap.Builder<String, String> responseHeaders = ImmutableListMultimap.builder();
```

Comment:
While it's possible, it will be quite complex:

```
        result.getStringHeaders().entrySet().stream()
                .flatMap(header -> header.getValue().stream()
                        .map(headerValue -> new Map.Entry(header.getKey(), headerValue)))
                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
```
(In addition, `new Map.Entry` doesn't actually work. I would need to come up with something else.)